File: PHP/One/Transforms/InferTypesPlugins/NullabilityCheckWithNot.php
<?php

namespace One\Transforms\InferTypesPlugins\NullabilityCheckWithNot;

use One\Transforms\InferTypesPlugins\Helpers\InferTypesPlugin\InferTypesPlugin;
use One\Ast\Expressions\Expression;
use One\Ast\Expressions\UnaryExpression;
use One\Ast\Expressions\BinaryExpression;
use One\Ast\Expressions\NullLiteral;
use One\Ast\AstTypes\ClassType;

class NullabilityCheckWithNot extends InferTypesPlugin {
    function __construct() {
        parent::__construct("NullabilityCheckWithNot");
        
    }
    
    function canTransform($expr) {
        return $expr instanceof UnaryExpression ? $expr->operator === "!" : false;
    }
    
    function transform($expr) {
        $unaryExpr = $expr;
        if ($unaryExpr->operator === "!") {
            $this->main->processExpression($expr);
            $type = $unaryExpr->operand->actualType;
            $litTypes = $this->main->currentFile->literalTypes;
            if ($type instanceof ClassType && $type->decl !== $litTypes->boolean->decl && $type->decl !== $litTypes->numeric->decl)
                return new BinaryExpression($unaryExpr->operand, "==", new NullLiteral());
        }
        
        return $expr;
    }
}

File: CSharp/One/Transforms/InferTypesPlugins/NullabilityCheckWithNot.cs
using One.Ast;
using One.Transforms.InferTypesPlugins.Helpers;

namespace One.Transforms.InferTypesPlugins
{
    public class NullabilityCheckWithNot : InferTypesPlugin {
        public NullabilityCheckWithNot(): base("NullabilityCheckWithNot")
        {
            
        }
        
        public override bool canTransform(Expression expr)
        {
            return expr is UnaryExpression unaryExpr ? unaryExpr.operator_ == "!" : false;
        }
        
        public override Expression transform(Expression expr)
        {
            var unaryExpr = ((UnaryExpression)expr);
            if (unaryExpr.operator_ == "!") {
                this.main.processExpression(expr);
                var type = unaryExpr.operand.actualType;
                var litTypes = this.main.currentFile.literalTypes;
                if (type is ClassType classType && classType.decl != litTypes.boolean.decl && classType.decl != litTypes.numeric.decl)
                    return new BinaryExpression(unaryExpr.operand, "==", new NullLiteral());
            }
            
            return expr;
        }
    }
}